spring/spring4/spring4-embedded/src/test/java/org/infinispan/spring/provider/SpringCacheCacheTest.java

```diff
@@ -212,6 +209,51 @@ public void testValueLoaderWithNoPreviousValue() {
       assertEquals("test", valueFromCache.get());
    }
 
+   /*
+    * In this test Thread 1 should exclusively block Cache#get method so that Thread 2 won't be able to
+    * insert "thread2" string into the cache.
+    *
+    * The test check this part of the Spring spec:
+    * Return the value to which this cache maps the specified key, obtaining that value from valueLoader if necessary.
+    * This method provides a simple substitute for the conventional "if cached, return; otherwise create, cache and return" pattern.
+    * @see http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/cache/Cache.html#get-java.lang.Object-java.util.concurrent.Callable-
+    */
+   @Test(timeOut = 30_000)
+   public void testValueLoaderWithLocking() throws Exception {
+      //given
+      CountDownLatch waitUntilThread1LocksValueGetter = new CountDownLatch(1);
+      CountDownLatch waitUntilBothThreadsFinish = new CountDownLatch(2);
+
+      //when
+      fork(() -> {
+         Thread.currentThread().setName("Thread1");
+         cache.get("test", () -> {
+            waitUntilThread1LocksValueGetter.countDown();
+            TimeUnit.MILLISECONDS.sleep(10);
+            return "thread1";
+         });
+         waitUntilBothThreadsFinish.countDown();
+      });
+
+      fork(() -> {
+         Thread.currentThread().setName("Thread2");
+         try {
+            waitUntilThread1LocksValueGetter.await();
+         } catch (InterruptedException e) {
+            throw new AssertionError(e);
+         }
+         cache.get("test", () -> "thread2");
```

Comment:
@slaskawi The question is what do you expect `cache.get("test", () -> "thread2")` to return. Should it return `thread1`, `thread2`, or it could be any of them?

If it could be any of them, then the test is fine, but if it should be `thread1`, then you need to assert that. Personally, I'd return the result of `cache.get(...)` from both `fork` lambdas, and then `assertEquals("thread1", thread1.get()); assertEquals("thread1", thread2.get());`

And if the implementation guarantees is that the lambda in thread 2 is never executed, then you should throw a `TestException` instead of returning `thread2`.